src/benchq/compilation/rbs_hyperparam_tuning.py

```diff
@@ -0,0 +1,451 @@
+    graph_data = empty_graph_re._get_graph_data_for_single_graph(graph_partition)
+
+    if space_or_time == "space":
+        return (10**time_overrun) + graph_data.max_graph_degree - 1
+    elif space_or_time == "time":
+        return (10**time_overrun) + graph_data.n_measurement_steps - 1
+    elif space_or_time == "space&time":
+        return (
+            (10**time_overrun) + graph_data.max_graph_degree - 1,
+            (10**time_overrun) + graph_data.n_measurement_steps - 1,
+        )
+    else:
+        raise ValueError(
+            "space_or_time must be 'space', 'time', or 'space&time'."
+            "Instead got " + str(space_or_time)
+        )
+
+
+def estimated_time_cost_from_rbs(
+    rbs_iteration_time: float,
+    circuit: Circuit,
+    verbose=False,
+    max_graph_size=None,
+    teleportation_threshold=40,
+    teleportation_distance=4,
+    min_neighbors=6,
+    max_num_neighbors_to_search=1e5,
+    decomposition_strategy=1,
+    circuit_prop_estimate=1.0,
+):
+    """
+    Runs Ruby Slippers (RBS) and uses the resulting graph to get estimate
+    of how long RBS would take to compile the whole circuit. This is a
+    low-level function. Unless you mean to use it, recommended to
+    use one of the functions beginning with "get_optimal_hyperparams_for_"
+
+    Args:
+        rbs_iteration_time (float): amount of time to let RBS run for
+        circuit (Circuit): the circuit to run RBS on
+        verbose (bool): true for verbose, false for quiet
+        max_graph_size (int): maximum number of nodes in the graph state
+        teleportation_threshold (int): RBS hyperparam, see ruby_slippers.jl
+        teleportation_distance (int): RBS hyperparam, see ruby_slippers.jl
+        min_neighbors (int): RBS hyperparam, see ruby_slippers.jl
+        max_num_neighbors_to_search (int): RBS hyperparam, see ruby_slippers.jl
+        decomposition_strategy (int): RBS hyperparam, see ruby_slippers.jl
+        circuit_prop_estimate (float): copying a large circuit to julia can take
+            a long time due to turning it into a string and back, so estimate how
+            much of the circuit RBS will get through before rbs_iteration_time
+            runs out, and supply that here to speed up optimization
+
+    Returns:
+        total estimated time to perform RBS on the whole circuit (float)
+    """
+    new_circuit = circuit
+    if circuit_prop_estimate != 1.0:
+        num_op = ceil(len(circuit.operations) * circuit_prop_estimate)
+        new_circuit = Circuit(circuit.operations[:num_op])
+
+    _, adj, iteration_prop = jl.run_ruby_slippers(
+        new_circuit,
+        verbose,
+        max_graph_size,
+        teleportation_threshold,
+        teleportation_distance,
+        min_neighbors,
+        max_num_neighbors_to_search,
+        decomposition_strategy,
+        rbs_iteration_time,
+    )
+
+    return rbs_iteration_time / (iteration_prop * circuit_prop_estimate)
```

Comment:
I'm probably just missing something, but why are we multiplying by `circuit_prop_estimate` here?

Reply:
I'm assuming that we care about the total time it would take RBS to get through 100% of the circuit. If the user supplies a `circuit_prop_estimate` that's != 1, the returned `proportion_of_circuit_completed` (`iteration_prop` above) will only be for that portion of the circuit, not the full circuit. Multiplying the denominator by the proportion of the circuit we pass to RBS (`circuit_prop_estimate`) gives the estimated time for 100% of the circuit

Reply:
OH! I see what's happening. Becase we are only passing part of the circuit the `proportion_of_circuit_completed` is just the proportion of that part of the circuit. So we need to compensate for that. Thanks for the clarification!